sled-storage/src/manager.rs

```diff
@@ -0,0 +1,969 @@
+                    if self.state == StorageManagerState::QueuingDisks =>
+                {
+                    if self.add_queued_disks().await {
+                        let _ = self.resource_updates.send_replace(self.resources.clone());
+                    }
+                }
+            }
+        }
+    }
+
+    /// Process the next event
+    ///
+    /// This is useful for testing/debugging
+    pub async fn step(&mut self) -> Result<(), Error> {
+        // The sending side should never disappear
+        let req = self.rx.recv().await.unwrap();
+        info!(self.log, "Received {:?}", req);
+        let should_send_updates = match req {
+            StorageRequest::AddDisk(raw_disk) => {
+                self.add_disk(raw_disk).await?
+            }
+            StorageRequest::RemoveDisk(raw_disk) => {
+                self.remove_disk(raw_disk).await
+            }
+            StorageRequest::DisksChanged(raw_disks) => {
+                self.ensure_using_exactly_these_disks(raw_disks).await
+            }
+            StorageRequest::NewFilesystem(request) => {
+                let result = self.add_dataset(&request).await;
+                if result.is_err() {
+                    warn!(self.log, "{result:?}");
+                }
+                let _ = request.responder.send(result);
+                false
+            }
+            StorageRequest::KeyManagerReady => {
+                self.state = StorageManagerState::Normal;
+                self.add_queued_disks().await
+            }
+            StorageRequest::GetLatestResources(tx) => {
+                let _ = tx.send(self.resources.clone());
+                false
+            }
+            StorageRequest::GetManagerState(tx) => {
+                let _ = tx.send(StorageManagerData {
+                    state: self.state,
+                    queued_u2_drives: self.queued_u2_drives.clone(),
+                });
+                false
+            }
+        };
+
+        if should_send_updates {
+            let _ = self.resource_updates.send_replace(self.resources.clone());
+        }
+
+        Ok(())
+    }
+
+    // Loop through all queued disks inserting them into [`StorageResources`]
+    // unless we hit a transient error. If we hit a transient error, we return
+    // and wait for the next retry window to re-call this method. If we hit a
+    // permanent error we log it, but we continue inserting queued disks.
+    //
+    // Return true if updates should be sent to watchers, false otherwise
+    async fn add_queued_disks(&mut self) -> bool {
+        info!(self.log, "Attempting to add queued disks");
+        self.state = StorageManagerState::Normal;
+
+        let mut send_updates = false;
+
+        // Disks that should be requeued.
+        let mut saved = HashSet::new();
+        let queued = std::mem::take(&mut self.queued_u2_drives);
+        let mut iter = queued.into_iter();
+        while let Some(disk) = iter.next() {
```

Comment:
Related to John's concerns about cancellation -- since this function is `&mut self` anyway, could we just "peek" each of the queued disks before  processing them, and only remove them if the processing was successful?

Doing so seems like it would simplify some of the logic around "who is responsible for re-queueing" -- we wouldn't need to juggle this `saved` set, since disks would only be removed from this queue once they've been processed successfully.

Reply:
This seemed like a reasonable idea, so I just started to implement it. However, it's trickier than it looks because you can't remove elements during iteration, and we can't use `retain` because the inside of the loop has async calls. We really want an `AsyncIterator` here.

Reply:
I just attempted another tactic to prevent cancellation issues. I'd track elements that need removing, but hanging onto the iterator while calling `self.add_u2_disk` runs into borrowck errors.

```
   async fn add_queued_disks(&mut self) -> bool {
         info!(self.log, "Attempting to add queued disks");
         self.state = StorageManagerState::Normal;

         let mut send_updates = false;

         // Disks that should be requeued.
●        let mut iter = self.queued_u2_drives.iter();
         let mut to_remove = HashSet::new();
●        while let Some(disk) = iter.next() {
             if self.state == StorageManagerState::QueueingDisks {
                 // We hit a transient error in a prior iteration.
                 break;
             } else {
                 // Try to add the disk. If there was a transient error the disk will
                 // have been requeued. If there was a permanent error, it will have been
                 // dropped. If there is an another unexpected error, we will handle it and
                 // requeue ourselves.
●                match self.add_u2_disk(disk.clone()).await {
                     Err(err) => {
                         warn!(
                             self.log,
                             "Potentially transient error: {err}: requeuing disk";
                             "disk_id" => ?disk.identity()
                         );
                     }
                     Ok(true) => {
                         send_updates = true;
                         to_remove.insert(disk);
                     }
                     Ok(false) => {
                         // There was no relevant change. The disk and its pool
                         // are already tracked in `StorageResources`;
                         to_remove.insert(disk);
                     }
                 }
             }
         }
         // Remove any processed disks
●        self.queued_u2_drives.retain(|k| !to_remove.contains(k));
         send_updates
     }
```

Reply:
I can work around the issue with iteration by cloning `queued_u2_drives` before iterating. I'll do that along with some other changes to `add_u2_disk` and this should be resolved.

Reply:
I fixed this up to be cancel safe in [11f1b36](https://github.com/oxidecomputer/omicron/pull/4332/commits/11f1b3695b4954e94ce40b9680e8e863b735c56f)